### DISCUSSION TITLE
Can I generate typescript entity from json?

### DISCUSSION THREAD

**Discussion by knight10-ux** [mikro-orm/mikro-orm#3846] [Q&A] [answered]
I want to store schema in the database and load entities dynamically, when needed. I see that with EntitySchema this could be possible but I have no way of initializing entity interface - in this case `IBook`.

```ts
export interface IBook extends CustomBaseEntity {
  title: string;
  author: Author;
  publisher?: Ref<Publisher>;
  tags: Collection<BookTag>;
}

export const Book = new EntitySchema<IBook, CustomBaseEntity>({
  name: 'Book',
  extends: 'CustomBaseEntity',
  properties: {
    title: { type: 'string' },
    author: { reference: 'm:1', entity: 'Author' },
    publisher: { reference: 'm:1', entity: 'Publisher', ref: true, nullable: true },
    tags: { reference: 'm:n', entity: 'BookTag', fixedOrder: true },
  },
});
```

What I would like to do is the following:

```ts
const bookObj = {
    title: { type: 'string' },
    author: { reference: 'm:1', entity: 'Author' },
    publisher: { reference: 'm:1', entity: 'Publisher', ref: true, nullable: true },
    tags: { reference: 'm:n', entity: 'BookTag', fixedOrder: true },
  }
export const Book = makeSchemaFromObj(bookObj);
```

Another reason for doing this is that I want to use composition instead of inheritance, because inheritance in my use case gets too complicated.

First I though about using library like `zod` and infer mikro-orm interface for `EntitySchema` but unfortunatelly I don't know how to handle Ref and Collections here.

Any help greatly appreciated!

**Comment by B4nan** [comment, marked as answer]
I don't understand your question. What do you mean by "I have no way of initializing entity interface - in this case IBook."? You can use `em.create()` for entities represented by interface only.

> What I would like to do is the following:

Your entity needs to have a name, otherwise its exactly the same as the previous (and working) example with EntitySchema.

**Reply by knight10-ux** [reply]
Thank you for quick answer!

> I don't understand your question.

My bad! I want to define entity using EntityShema class without providing interface - just properties (in future handling indexes, validators as well). Is it possible to infer `IBook` interface from obj?

```ts
const bookObj = {
    title: { type: 'string' },
    author: { reference: 'm:1', entity: 'Author' },
    publisher: { reference: 'm:1', entity: 'Publisher', ref: true, nullable: true },
    tags: { reference: 'm:n', entity: 'BookTag', fixedOrder: true },
  }
export const Book = makeSchemaFromObj('Book', bookObj);
```

Book will have publisher prop of type "Ref\<Publisher\>" here.

What I really want to do is the following:

```ts
const ownedComponent = {
  owner: { reference: 'm:1', entity: 'Author' },
  createdAt: { type: Date, onCreate: () => new Date() },
}

const titleComponent = {
  title: { type: 'string' },
  slug: { type: 'string' },
}

const tagComponent = {
  tags: { reference: 'm:n', entity: 'Tag', fixedOrder: true },
}

export const Book = makeSchemaFromObj('Book', [ownedComponent, titleComponent, tagComponent]);
```

**Reply by B4nan** [reply]
>  Is it possible to infer IBook interface from obj?

Nope, but I am sure you can use zod for that. Can't really help with that, never used it myself. 

But it's true we could have a way to infer it, right now it works in the other direction - it validates your interface matches the definition. Keep in mind the `EntitySchema` instance is a runtime object, it's not just a typescript type. There would have to be some type helper to infer it from the schema instance. I will think about this for v6, but a zod interop would be even better, given how popular that library is.

> Book will have publisher prop of type "Ref<Publisher>" here.

Yes, because you have the `ref: true` in that property definition, that is an optional feature.

**Reply by knight10-ux** [reply]
Lovely, thanks again for the reply and have a good day!
